evm_arithmetization/src/curve_pairings.rs

```diff
@@ -150,366 +141,937 @@ where
+            im: F::ZERO,
         },
         t1: g01,
-        t2: Fp2::<BN254>::ZERO,
+        t2: Fp2::<F>::ZERO,
     };
 
     let g1 = Fp6 {
-        t0: Fp2::<BN254>::ZERO,
+        t0: Fp2::<F>::ZERO,
         t1: g11,
-        t2: Fp2::<BN254>::ZERO,
+        t2: Fp2::<F>::ZERO,
     };
 
     Fp12 { z0: g0, z1: g1 }
 }
 
-pub(crate) fn gen_bn_fp12_sparse<R: Rng + ?Sized>(rng: &mut R) -> Fp12<BN254> {
-    bn_sparse_embed(
-        rng.gen::<BN254>(),
-        rng.gen::<Fp2<BN254>>(),
-        rng.gen::<Fp2<BN254>>(),
-    )
+pub(crate) fn check_curve_eq_aff<T>(p: CurveAff<T>, b_coeff: T) -> bool
+where
+    T: FieldExt,
+{
+    p.y * p.y == p.x * p.x * p.x + b_coeff
 }
 
-/// The output y of the miller loop is not an invariant,
-/// but one gets an invariant by raising y to the power
-///     (p^12 - 1)/N = (p^6 - 1)(p^2 + 1)(p^4 - p^2 + 1)/N
-/// where N is the cyclic group order of the curve.
-/// To achieve this, we first exponentiate y by p^6 - 1 via
-///     y = y_6 / y
-/// and then exponentiate the result by p^2 + 1 via
-///     y = y_2 * y
-/// We then note that (p^4 - p^2 + 1)/N can be rewritten as
-///     (p^4 - p^2 + 1)/N = p^3 + (a2)p^2 - (a1)p - a0
-/// where 0 < a0, a1, a2 < p. Then the final power is given by
-///     y = y_3 * (y^a2)_2 * (y^-a1)_1 * (y^-a0)
-pub(crate) fn bn_final_exponent(f: Fp12<BN254>) -> Fp12<BN254> {
-    let mut y = f.frob(6) / f;
-    y = y.frob(2) * y;
-    let (y_a2, y_a1, y_a0) = get_bn_custom_powers(y);
-    y.frob(3) * y_a2.frob(2) * y_a1.frob(1) * y_a0
+pub(crate) fn check_curve_eq_proj<T>(p: CurveProj<T>, b_coeff: T) -> bool
+where
+    T: FieldExt,
+{
+    p.y * p.y == p.x * p.x * p.x + b_coeff
 }
 
-/// We first together (so as to avoid repeated steps) compute
-///     y^a4, y^a2, y^a0
-/// where a1 is given by
-///     a1 = a4 + 2a2 - a0
-/// we then invert y^a0 and return
-///     y^a2, y^a1 = y^a4 * y^a2 * y^a2 * y^(-a0), y^(-a0)
-///
-/// Representing a4, a2, a0 in *little endian* binary, define
-///     BN_EXPS4 = [(a4[i], a2[i], a0[i]) for i in       0..len(a4)]
-///     BN_EXPS2 = [       (a2[i], a0[i]) for i in len(a4)..len(a2)]
-///     BN_EXPS0 = [               a0[i]  for i in len(a2)..len(a0)]
-fn get_bn_custom_powers(f: Fp12<BN254>) -> (Fp12<BN254>, Fp12<BN254>, Fp12<BN254>) {
-    let mut sq: Fp12<BN254> = f;
-    let mut y0: Fp12<BN254> = Fp12::<BN254>::UNIT;
-    let mut y2: Fp12<BN254> = Fp12::<BN254>::UNIT;
-    let mut y4: Fp12<BN254> = Fp12::<BN254>::UNIT;
-
-    // proceed via standard squaring algorithm for exponentiation
-
-    // must keep multiplying all three values: a4, a2, a0
-    for (a, b, c) in BN_EXPS4 {
-        if a {
-            y4 = y4 * sq;
+/// Generates a sparse, random Fp12 element.
+pub(crate) fn gen_fp12_sparse<F, R: Rng + ?Sized>(rng: &mut R) -> Fp12<F>
+where
+    F: FieldExt,
+    Fp2<F>: Adj,
+    Standard: Distribution<F>,
+{
+    sparse_embed::<F>(rng.gen::<F>(), rng.gen::<Fp2<F>>(), rng.gen::<Fp2<F>>())
+}
+
+pub mod bn254 {
```

Comment:
Because we need the exact same implementation as in the kernel to test different methods (and also to not have to rely on additional dependencies). I didn't touch anything there, just moved inside a new module.